llm_studio/src/models/text_causal_language_modeling_model.py

```diff
@@ -142,66 +220,259 @@ def generate(self, batch: Dict, cfg: Any):
             [
                 TokenStoppingCriteria(
                     stop_word_ids=self.cfg.tokenizer._stop_words_ids,
-                    prompt_input_ids_len=batch["prompt_input_ids"].shape[1],
+                    prompt_input_ids_len=input_ids.shape[1],
                 )
             ]
         )
 
+        # The KL-div estimation assumes sampling and specific settings
+        if self.training and cfg.training.use_rlhf:
+            do_sample = True
+            temperature = cfg.training.ppo_generate_temperature
+            top_k = 0.0
+            top_p = 1.0
+            repetition_penalty = 1.0
+        else:
+            do_sample = cfg.prediction.do_sample
+            temperature = float(cfg.prediction.temperature)
+            top_k = cfg.prediction.top_k
+            top_p = float(cfg.prediction.top_p)
+            repetition_penalty = float(cfg.prediction.repetition_penalty)
+
         transformers_logging.set_verbosity_error()
         output = generation_function(
-            inputs=batch["prompt_input_ids"],
-            attention_mask=batch["prompt_attention_mask"],
+            inputs=input_ids,
+            attention_mask=attention_mask,
             pad_token_id=pad_token_id,
             min_new_tokens=cfg.prediction.min_length_inference,
             max_new_tokens=cfg.prediction.max_length_inference,
-            do_sample=cfg.prediction.do_sample,
+            do_sample=do_sample,
             num_beams=cfg.prediction.num_beams,
-            temperature=float(cfg.prediction.temperature),
-            repetition_penalty=float(cfg.prediction.repetition_penalty),
-            top_k=cfg.prediction.top_k,
-            top_p=float(cfg.prediction.top_p),
+            temperature=temperature,
+            repetition_penalty=repetition_penalty,
+            top_k=top_k,
+            top_p=top_p,
             stopping_criteria=stopping_criteria,
             renormalize_logits=True,
             return_dict_in_generate=False,
             use_cache=True,
         )
         transformers_logging.set_verbosity(verbosity)
 
-        # Mask the prompt tokens
-        output[:, : batch["prompt_input_ids"].shape[1]] = pad_token_id
+        if remove_prompt:
+            # remove the prompt tokens
+            output = output[:, input_ids.shape[1] :]
+        else:
+            # Mask the prompt tokens
+            output[:, : input_ids.shape[1]] = pad_token_id
 
         return output
 
     def forward(
         self,
         batch: Dict,
-        calculate_loss: bool = True,
+        generate: bool = False,
+        padding: bool = True,
+        is_ref_model: bool = False,
```

Comment:
I think one can remove `is_ref_model` flag and also compute loss/value in the reference model (they will not be used afterwards, of course). This would result in slightly slimmer code. WDYT?